### PR TITLE
Fix `PyTorchYolo` Incorrect Inputs and Unify Input Processing

### LINKED ISSUE #2088: Incorrect YOLO Bounding Box Input Format
**Describe the bug**
The YOLOv3 and YOLOv5 estimators require bounding boxes to be in the normalized xywh format as indicated the [YOLOv3 docs](https://github.com/eriklindernoren/PyTorch-YOLOv3/tree/master#train-on-custom-dataset) and [YOLOv5 docs](https://docs.ultralytics.com/yolov5/train_custom_data/#12-create-labels_1). The current `PyTorchYolo` estimator uses the `translate_labels_art_to_yolov3` which does not perform the normalization. Therefore, the inputs for the `compute_loss` and `fit` methods are invalid.

**To Reproduce**
The following sample code can reproduce the issue:

```python
import torch
from art.estimators.object_detection.pytorch_yolo import translate_labels_art_to_yolov3

y = [{
    'boxes': torch.tensor([
        [144., 185., 952., 709.],
        [676., 44., 1150., 691.],
        [429., 433., 517., 704.],
        [979., 316., 1022., 419.],
    ]),
    'labels': torch.tensor([0, 0, 27, 27])
}]
```

If running the `translate_labels_art_to_yolov3`, the output is:

```python
translate_labels_art_to_yolov3(y)
```

The output is:

```txt
tensor([[  0.,   0., 144., 185., 808., 524.],
        [  0.,   0., 676.,  44., 474., 647.],
        [  0.,  27., 429., 433.,  88., 271.],
        [  0.,  27., 979., 316.,  43., 103.]])
```

This is not normalized, the correct format should be normalized like this:

```txt
tensor([[ 0.0000,  0.0000,  0.6850,  0.5588,  1.0100,  0.6550],
        [ 0.0000,  0.0000,  1.1413,  0.4594,  0.5925,  0.8087],
        [ 0.0000, 27.0000,  0.5913,  0.7106,  0.1100,  0.3388],
        [ 0.0000, 27.0000,  1.2506,  0.4594,  0.0538,  0.1287])
```

**Expected behavior**
The `translate_labels_art_to_yolov3` should be fixed to normalize the bounding boxes and produce the correct inputs in the normalized xywh format as indicated by the docs.

**Screenshots**
N/A

**System information (please complete the following information):**
 - OS
 - Python version
 - ART version or commit number
 - TensorFlow / Keras / PyTorch / MXNet version

### PR DESCRIPTION
**Pull request by f4str**
# Description

Replace the invalid `translate_labels_art_to_yolov3` function with `translate_labels_x1y1x2y2_to_xcycwh` which correctly normalizes the bounding boxes for the YOLO inputs (and it more descriptive). Additionally clean up the `translate_predictions_xcycwh_to_x1y1x2y2` such that it is easier to understand. **There was also another bug found where this converted input is not put on the GPU. This bug was also addressed in this PR.**

In order to ensure all inputs are properly formatted, a new method `_preprocess_and_convert_inputs` was created. All of the methods (`fit`, `predict`, `compute_loss`, etc.) now use this new method to apply preprocessing and convert inputs to the proper tensor format. This not only makes a cleaner implementation, but as a consequence causes the `channels_first` parameter to work fully for all methods in the `PyTorchYolo` estimator.

Lastly, update the docs to include the missing PyTorch YOLO estimator.

Fixes #2088 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Updated tests for PyTorchYolo

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/object_detection/pytorch_yolo.py [hunk 1]
```diff
@@ -43,74 +43,78 @@ def translate_predictions_xcycwh_to_x1y1x2y2(
...
 ) -> List[Dict[str, "torch.Tensor"]]:
     """
-    Convert object detection predictions from xcycwh to x1y1x2y2 format.
+    Convert object detection predictions from xcycwh (YOLO) to x1y1x2y2 (torchvision).
 
-    :param y_pred_xcycwh: Labels in format xcycwh.
-    :return: Labels in format x1y1x2y2.
+    :param y_pred_xcycwh: Object detection labels in format xcycwh (YOLO).
+    :param height: Height of images in pixels.
+    :param width: Width if images in pixels.
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:50]:
```suggestion
    :param input_height: Height of images in pixels.
    :param input_width: Width if images in pixels.
```

File: art/estimators/object_detection/pytorch_yolo.py [hunk 2]
```diff
@@ -43,74 +43,78 @@ def translate_predictions_xcycwh_to_x1y1x2y2(
...
+) -> "torch.Tensor":
     """
-    Translate labels from ART to YOLO v3 and v5.
+    Translate object detection labels from x1y1x2y2 (torchvision) to xcycwh (YOLO).
 
-    :param labels_art: Object detection labels in format ART (torchvision).
-    :return: Object detection labels in format YOLO v3 and v5.
+    :param labels_x1y1x2y2: Object detection labels in format x1y1x2y2 (torchvision).
+    :param height: Height of images in pixels.
+    :param width: Width if images in pixels.
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:89]:
```suggestion
    :param input_height: Height of images in pixels.
    :param input_width: Width if images in pixels.
```

File: art/estimators/object_detection/pytorch_yolo.py [hunk 3]
```diff
@@ -503,44 +535,24 @@ def fit(  # pylint: disable=W0221
...
-        # Convert samples into tensors
-        transform = torchvision.transforms.Compose([torchvision.transforms.ToTensor()])
-
-        if self.clip_values is not None:
-            norm_factor = self.clip_values[1]
-        else:
-            norm_factor = 1.0
+        # Apply preprocessing and convert to tensors
+        x_preprocessed, y_preprocessed_list = self._preprocess_and_convert_inputs(x=x, y=y, fit=True, no_grad=True)
+        # Cast to np.ndarray to use list indexing
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:547]:
```suggestion
        x_preprocessed, y_preprocessed_list = self._preprocess_and_convert_inputs(x=x, y=y, fit=True, no_grad=True)

        # Cast to np.ndarray to use list indexing
```

File: art/estimators/object_detection/pytorch_yolo.py [hunk 4]
```diff
@@ -256,149 +267,186 @@ def device(self) -> "torch.device":
...
+        :param fit: `True` if the function is call before fit/training and `False` if the function is called before a
+                    predict operation.
+        :param no_grad: `True` if no gradients required.
+        :return: Preprocessed inputs `(x, y)` as tensors.
         """
         import torch
 
-        self._model.train()
-        self.set_batchnorm(train=False)
-        self.set_dropout(train=False)
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:278]: Where are  `set_batchnorm` and `set_dropout` applied now?

I think we might have set them in `loss_gradient`, `predict` and `fit` to make sure these layers are set correctly.

File: tests/estimators/object_detection/test_pytorch_yolo.py [hunk 5]
```diff
@@ -173,116 +173,116 @@ def test_loss_gradient(art_warning, get_pytorch_yolo):
...
+                -0.005740379,
+                0.009537468,
+                -0.01116704,
+                0.0010225883,
+                0.00026052812,
             ]
         )
 
-        np.testing.assert_array_almost_equal(grads[0, 0, 208, 175:225], expected_gradients1, decimal=2)
+        np.testing.assert_array_almost_equal(grads[0, 0, 208, 175:225], expected_gradients1, decimal=1)
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_pytorch_yolo.py:229]: Is the precision change required?

File: tests/estimators/object_detection/test_pytorch_yolo.py [hunk 6]
```diff
@@ -173,116 +173,116 @@ def test_loss_gradient(art_warning, get_pytorch_yolo):
...
+                0.03357561,
+                0.01510548,
+                0.0006220075,
+                -0.027102726,
+                -0.020182101,
+                -0.04347762,
             ]
         )
-        np.testing.assert_array_almost_equal(grads[1, 0, 208, 175:225], expected_gradients2, decimal=2)
+        np.testing.assert_array_almost_equal(grads[1, 0, 208, 175:225], expected_gradients2, decimal=1)
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_pytorch_yolo.py:285]: Is the precision change required?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @f4str Thank you for fixing this issue! I have added a question about the functions setting train mode in dropout and batchnorm layers and a few minor suggestions. What do you think?

File: tests/estimators/object_detection/test_pytorch_yolo.py (hunk 5 above)
Comment (f4str) [reply tests/estimators/object_detection/test_pytorch_yolo.py:229]: Due to normalizing the inputs, the gradients have now changed in magnitude and are much larger now. Therefore, the decimal precision needs to be decreased.

File: tests/estimators/object_detection/test_pytorch_yolo.py (hunk 6 above)
Comment (f4str) [reply tests/estimators/object_detection/test_pytorch_yolo.py:285]: Due to normalizing the inputs, the gradients have now changed in magnitude and are much larger now. Therefore, the decimal precision needs to be decreased.

File: art/estimators/object_detection/pytorch_yolo.py (hunk 4 above)
Comment (f4str) [reply art/estimators/object_detection/pytorch_yolo.py:278]: Based on our discussion, the `set_batchnorm` and `set_dropout` should no longer be needed as subsequent calls to the `loss_gradient` yield gradients that are still the same within numerical accuracy.

File: art/estimators/object_detection/pytorch_yolo.py (hunk 1 above)
Comment (f4str) [reply art/estimators/object_detection/pytorch_yolo.py:50]: I think it actually makes more sense to rename the variable than the docstring. The current names `input_height` and `input_width` are slightly redundant since it should be clear that they refer to the input image.

File: art/estimators/object_detection/pytorch_yolo.py (hunk 2 above)
Comment (f4str) [reply art/estimators/object_detection/pytorch_yolo.py:89]: I think it actually makes more sense to rename the variable than the docstring. The current names `input_height` and `input_width` are slightly redundant since it should be clear that they refer to the input image.

**Comment by f4str** [comment]
Hi @beat-buesser thank you for the suggestions. I've replied to the comments and have also made some corresponding changes. Feel free to add any additional comments.

**Review by beat-buesser — APPROVED** [review]
